File: hoffman-agents-python/src/conscious_agent/core/experience_trie.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Self


@dataclass
class TrieNode:
    state_id: int
    children: dict[int, TrieNode] = field(default_factory=dict)
    visit_count: int = 0
    prediction_errors: list[float] = field(default_factory=list)
    mean_prediction_error: float = 0.0
    word_binding: str | None = None
    depth: int = 0

    def add_prediction_error(self, error: float) -> None:
        n = len(self.prediction_errors)
        self.mean_prediction_error = (self.mean_prediction_error * n + error) / (n + 1)
        self.prediction_errors.append(error)


class ExperienceTrie:
    def __init__(self, max_depth: int = 10) -> None:
        if max_depth < 1:
            raise ValueError(f"max_depth must be >= 1, got {max_depth}")
        self._root = TrieNode(state_id=-1)
        self._max_depth = max_depth

    def insert(self, path: list[int], prediction_error: float = 0.0) -> None:
        if not path:
            return
        if len(path) > self._max_depth:
            path = path[:self._max_depth]
        node = self._root
        for depth, state_id in enumerate(path, start=1):
            if state_id not in node.children:
                node.children[state_id] = TrieNode(state_id=state_id, depth=depth)
            node = node.children[state_id]
            node.visit_count += 1
        node.add_prediction_error(prediction_error)

    def lookup(self, path: list[int]) -> TrieNode | None:
        node = self._root
        for state_id in path:
            if state_id not in node.children:
                return None
            node = node.children[state_id]
        return node

    def predict_next(self, path: list[int]) -> int | None:
        node = self.lookup(path)
        if node is None or not node.children:
            return None
        return max(node.children.items(), key=lambda x: x[1].visit_count)[0]

    def get_all_paths(self, min_visits: int = 1) -> list[list[int]]:
        paths: list[list[int]] = []

        def dfs(node: TrieNode, current: list[int]) -> None:
            if node is not self._root and node.visit_count >= min_visits:
                paths.append(list(current))
            for child_state, child in node.children.items():
                current.append(child_state)
                dfs(child, current)
                current.pop()

        dfs(self._root, [])
        return paths

    def get_leaf_paths(self, min_visits: int = 1) -> list[list[int]]:
        paths: list[list[int]] = []

        def dfs(node: TrieNode, current: list[int]) -> None:
            if not node.children:
                if node is not self._root and node.visit_count >= min_visits:
                    paths.append(list(current))
                return
            for child_state, child in node.children.items():
                current.append(child_state)
                dfs(child, current)
                current.pop()

        dfs(self._root, [])
        return paths

    def compress(self, min_visits: int) -> None:
        def _prune(node: TrieNode) -> bool:
            children_to_delete: list[int] = []
            for child_state, child in node.children.items():
                if _prune(child):
                    children_to_delete.append(child_state)
            for cs in children_to_delete:
                del node.children[cs]
            if node is not self._root and node.visit_count < min_visits and not node.children:
                return True
            return False

        _prune(self._root)

    def merge(self, other: ExperienceTrie) -> ExperienceTrie:
        merged = ExperienceTrie(max_depth=max(self._max_depth, other._max_depth))

        path_data: dict[tuple[int, ...], tuple[int, list[float]]] = {}

        def _collect(node: TrieNode, path: list[int]) -> None:
            for child_state, child in node.children.items():
                path.append(child_state)
                key = tuple(path)
                vc, errs = path_data.get(key, (0, []))
                path_data[key] = (vc + child.visit_count, errs + child.prediction_errors)
                _collect(child, path)
                path.pop()

        _collect(self._root, [])
        _collect(other._root, [])

        for path_tuple in sorted(path_data.keys(), key=len):
            path = list(path_tuple)
            visit_count, pred_errors = path_data[path_tuple]
            merged.insert(path)
            node = merged.lookup(path)
            if node is not None:
                node.visit_count = visit_count
                node.prediction_errors = pred_errors
                if pred_errors:
                    node.mean_prediction_error = sum(pred_errors) / len(pred_errors)

        return merged

    def size(self) -> int:
        count = 0

        def _count(node: TrieNode) -> None:
            nonlocal count
            count += 1
            for child in node.children.values():
                _count(child)

        _count(self._root)
        return count

    def compression_ratio(self, total_steps: int) -> float:
        if total_steps == 0:
            return 0.0
        return self.size() / total_steps

    @property
    def root(self) -> TrieNode:
        return self._root

    @property
    def max_depth(self) -> int:
        return self._max_depth

    def clear(self) -> None:
        self._root = TrieNode(state_id=-1)

File: hoffman-agents-react-native/src/core/experience-trie.js
class TrieNode {
  constructor(stateId, depth = 0) {
    this.stateId = stateId;
    this.children = {};
    this.visitCount = 0;
    this.predictionErrors = [];
    this.meanPredictionError = 0;
    this.wordBinding = null;
    this.depth = depth;
  }

  addPredictionError(error) {
    const n = this.predictionErrors.length;
    this.meanPredictionError = (this.meanPredictionError * n + error) / (n + 1);
    this.predictionErrors.push(error);
  }
}

class ExperienceTrie {
  constructor(maxDepth = 10) {
    if (maxDepth < 1) throw new Error(`max_depth must be >= 1, got ${maxDepth}`);
    this._root = new TrieNode(-1);
    this._maxDepth = maxDepth;
  }

  insert(path, predictionError = 0) {
    if (!path || path.length === 0) return;
    if (path.length > this._maxDepth) path = path.slice(0, this._maxDepth);
    let node = this._root;
    for (let depth = 0; depth < path.length; depth++) {
      const stateId = path[depth];
      if (!node.children[stateId]) {
        node.children[stateId] = new TrieNode(stateId, depth + 1);
      }
      node = node.children[stateId];
      node.visitCount++;
    }
    node.addPredictionError(predictionError);
  }

  lookup(path) {
    let node = this._root;
    for (const stateId of path) {
      if (!node.children[stateId]) return null;
      node = node.children[stateId];
    }
    return node;
  }

  predictNext(path) {
    const node = this.lookup(path);
    if (!node || Object.keys(node.children).length === 0) return null;
    let bestState = null;
    let bestCount = -1;
    for (const [stateId, child] of Object.entries(node.children)) {
      if (child.visitCount > bestCount) {
        bestCount = child.visitCount;
        bestState = parseInt(stateId);
      }
    }
    return bestState;
  }

  _dfsPaths(node, current, minVisits, results, leavesOnly) {
    if (node !== this._root && node.visitCount >= minVisits) {
      if (!leavesOnly || Object.keys(node.children).length === 0) {
        results.push([...current]);
      }
    }
    for (const [childState, child] of Object.entries(node.children)) {
      current.push(parseInt(childState));
      this._dfsPaths(child, current, minVisits, results, leavesOnly);
      current.pop();
    }
  }

  getAllPaths(minVisits = 1) {
    const results = [];
    this._dfsPaths(this._root, [], minVisits, results, false);
    return results;
  }

  getLeafPaths(minVisits = 1) {
    const results = [];
    this._dfsPaths(this._root, [], minVisits, results, true);
    return results;
  }

  compress(minVisits) {
    const _prune = (node) => {
      const toDelete = [];
      for (const [cs, child] of Object.entries(node.children)) {
        if (_prune(child)) toDelete.push(parseInt(cs));
      }
      for (const cs of toDelete) delete node.children[cs];
      if (node !== this._root && node.visitCount < minVisits && Object.keys(node.children).length === 0) {
        return true;
      }
      return false;
    };
    _prune(this._root);
  }

  merge(other) {
    const merged = new ExperienceTrie(Math.max(this._maxDepth, other._maxDepth));
    const pathData = new Map();

    const collect = (node, path) => {
      for (const [cs, child] of Object.entries(node.children)) {
        path.push(parseInt(cs));
        const key = path.join(',');
        const existing = pathData.get(key) || [0, []];
        existing[0] += child.visitCount;
        existing[1] = existing[1].concat(child.predictionErrors);
        pathData.set(key, existing);
        collect(child, path);
        path.pop();
      }
    };

    collect(this._root, []);
    collect(other._root, []);

    for (const [keyStr, [visitCount, predErrors]] of pathData) {
      const path = keyStr.split(',').map(Number);
      merged.insert(path);
      const node = merged.lookup(path);
      if (node) {
        node.visitCount = visitCount;
        node.predictionErrors = predErrors;
        node.meanPredictionError = predErrors.length > 0
          ? predErrors.reduce((a, b) => a + b, 0) / predErrors.length
          : 0;
      }
    }
    return merged;
  }

  _collectPaths(node, currentPath, minVisits, results) {
    if (node !== this._root && node.visitCount >= minVisits) {
      results.push({ path: [...currentPath], visitCount: node.visitCount, meanPredictionError: node.meanPredictionError, wordBinding: node.wordBinding });
    }
    for (const [stateId, child] of Object.entries(node.children)) {
      currentPath.push(parseInt(stateId));
      this._collectPaths(child, currentPath, minVisits, results);
      currentPath.pop();
    }
  }

  getStats() {
    let nodeCount = 0, maxDepth = 0, totalVisits = 0, visitSum = 0;
    const depthDist = {};
    const walk = (node, depth) => {
      nodeCount++;
      maxDepth = Math.max(maxDepth, depth);
      totalVisits += node.visitCount;
      visitSum += node.visitCount;
      depthDist[depth] = (depthDist[depth] || 0) + 1;
      for (const child of Object.values(node.children)) walk(child, depth + 1);
    };
    walk(this._root, 0);
    return { nodeCount, maxDepth, totalVisits, meanVisitCount: nodeCount > 0 ? totalVisits / nodeCount : 0, depthDistribution: depthDist };
  }

  exportNodes(minVisits = 1) {
    const results = [];
    this._collectPaths(this._root, [], minVisits, results);
    return results;
  }

  getDominantPaths(topK = 10) {
    const all = this.exportNodes(1);
    all.sort((a, b) => b.visitCount - a.visitCount);
    return all.slice(0, topK);
  }

  size() {
    let count = 0;
    const countNodes = (node) => {
      count++;
      for (const child of Object.values(node.children)) countNodes(child);
    };
    countNodes(this._root);
    return count;
  }

  compressionRatio(totalSteps) {
    if (totalSteps === 0) return 0;
    return this.size() / totalSteps;
  }

  get root() { return this._root; }
  get maxDepth() { return this._maxDepth; }

  clear() {
    this._root = new TrieNode(-1);
  }
}

module.exports = { ExperienceTrie, TrieNode };
